src/components/Messages/index.tsx

```diff
@@ -29,67 +69,63 @@ const Messages: FC = () => {
-    if (!client || !isFeatureEnabled('messages', currentProfile?.id)) {
+    if (!isMessagesEnabled || !client) {
       return;
     }
 
-    async function listConversations() {
-      setLoading(true);
+    const fetchMostRecentMessage = async (
+      convo: Conversation
+    ): Promise<{ address: string; preview: MessagePreview }> => {
+      const peerAddress = convo.peerAddress.toLowerCase();
+      const newMessagePreview = new MessagePreview(messagePreviews.get(peerAddress));
+      // TODO(elise): Add sort direction on XMTP's side so we can grab only the most recent message.
+      const newMessages = await convo.messages({ limit: 1 });
+      if (newMessages.length >= 0) {
+        newMessagePreview.message = newMessages[0];
+      }
+      return { address: peerAddress, preview: newMessagePreview };
+    };
+
+    const listConversations = async () => {
+      const newMessagePreviews = new Map(messagePreviews);
+      const newConversations = new Map(conversations);
       const convos = (await client?.conversations?.list()) || [];
-      Promise.all(
+      const previews = await Promise.all(
         convos.map(async (convo) => {
-          if (convo.peerAddress !== currentProfile?.ownedBy) {
-            conversations.set(convo.peerAddress, convo);
-            setConversations(new Map(conversations));
-          }
+          newConversations.set(convo.peerAddress, convo);
+          return await fetchMostRecentMessage(convo);
         })
-      ).then(() => {
-        setLoading(false);
-      });
-    }
-    const streamConversations = async () => {
-      const newStream = (await client?.conversations?.stream()) || [];
-      setStream(newStream);
-      for await (const convo of newStream) {
-        if (convo.peerAddress !== currentProfile?.ownedBy) {
-          const newMessages = await convo.messages();
-          messages.set(convo.peerAddress, newMessages);
-          setMessages(new Map(messages));
-          conversations.set(convo.peerAddress, convo);
-          setConversations(new Map(conversations));
-        }
+      );
+      for (const preview of previews) {
+        newMessagePreviews.set(preview.address, preview.preview);
       }
+      setMessagePreviews(newMessagePreviews);
+      setConversations(newConversations);
     };
+
     listConversations();
-    streamConversations();
-
-    return () => {
-      const closeStream = async () => {
-        if (!stream) {
-          return;
-        }
-        await stream.return();
-      };
-      closeStream();
-    };
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [client]);
 
-  const onConversationSelected = (address: string) => {
-    router.push(address ? `/messages/${address}` : '/messages/');
-  };
-
-  if (!isFeatureEnabled('messages', currentProfile?.id)) {
+  if (!isMessagesEnabled) {
     return <Custom404 />;
   }
 
+  if (profilesError) {
+    return <Custom500 />;
+  }
+
+  if (!messagePreviews) {
+    return null;
```

Comment:
We should return something like a loader here, and if absolutely no conversations / messages exist we should have a no message / no conversation UI to be shown

Reply:
100% agree! I actually had it in this [PR](https://github.com/lensterxyz/lenster/pull/779/commits/52db08a1d885d79e1d7103e614564b4936355f87#diff-5635ed0d7bc771d451cd29937e15f80549c88f309500bde6f80ed7a282bbe5d9L153), but removed it because it was showing when the data was available and I realized it would be easier to add when I pull this out to its own hook.

Tracking here: https://github.com/xmtp-labs/hq/issues/735